guides/inventing-service.md

```diff
@@ -0,0 +1,856 @@
+We can wrap `handler_with_timeout` in a similar way and modify the response like
+so:
+
+```rust
+async fn handler_with_timeout_and_content_type(
+    request: Request,
+) -> Result<Response, Error> {
+    let mut response = handler_with_timeout(request).await?;
+    response.set_header("Content-Type", "application/json");
+    Ok(response)
+}
+```
+
+We now have a handler that will process an HTTP request, take no longer than 30
+seconds, and always have the right `Content-Type` header all without modifying
+our original `handle_request` function or `Server` struct.
+
+Designing libraries that can be extended in this way is very powerful since it
+allows users to extend things without having to wait for the library maintainers
+to add support for it.
+
+However there is one problem. Our current setup doesn't scale very well. Imagine
+we have many `handle_with_*` functions that each add a little bit of behavior.
+Having to hard code the chain of which intermediate handler calls which isn't
+great. Our current chain is
+
+1. `handler_with_timeout_and_content_type` which calls
+2. `handler_with_timeout` which calls
+3. `handle_request` which actually processes the request
+
+It would be nice if we could somehow compose these three functions without
+having to hard code the exact order. Something like:
+
+```rust
+let final_handler = with_content_type(with_timeout(handle_request));
+```
+
+While still being able to run our handler like before:
+
+```rust
+server.run(final_handler).await?;
+```
+
+You could imagine `with_content_type` and `with_timeout` being functions that
+took an argument of type `F: Fn(Request) -> Future<Output = Response>` and
+returned a closure a la `impl Fn(Request) -> Future<Output = Result<Response,
+Error>>` but all these closure types quickly become hard to deal with.
+
+## The `Handler` trait
+
+Lets try another approach. Rather than `Server::run` accepting a closure
+(`Fn(Request) -> ...`) lets make a new trait that encapsulates the same `async
+fn(Request) -> Result<Response, Error>` setup:
+
+```rust
+trait Handler {
+    async fn call(&mut self, request: Request) -> Result<Response, Error>;
+}
+```
+
+Having a trait like this allows us to write concrete types that implement it so
+we don't have to deal with `Fn`s all the time.
+
+However Rust currently doesn't support async traits so we have two options:
+
+1. Make `call` return a boxed future like `Pin<Box<dyn Future<Output =
+   Result<Response, Error>>>`. This is what the [async-trait] crate does.
+2. Add an associated `type Future` to `Handler` so users get to pick their own
+   type.
+
+Lets go with option 2 as its the most flexible. Users who have a concrete future
+type can use that without the cost of a `Box` and users who don't care can still
+use `Pin<Box<...>>`.
+
+```rust
+trait Handler {
+    type Future: Future<Output = Result<Response, Error>>;
+
+    fn call(&mut self, request: Request) -> Self::Future;
```

Comment:
I would also like to add a paragraph about why `call` taking `&mut self` is necessary. As I don't think thats entirely obvious.